optuna/study/_frozen.py

```diff
@@ -0,0 +1,92 @@
+from typing import Any
+from typing import Dict
+from typing import Optional
+from typing import Sequence
+
+from optuna import logging
+from optuna.study._study_direction import StudyDirection
+
+
+_logger = logging.get_logger(__name__)
+
+
+class FrozenStudy:
+    """Basic attributes of a :class:`~optuna.study.Study`.
+
+    Attributes:
+        study_name:
+            Name of the :class:`~optuna.study.Study`.
+        direction:
+            :class:`~optuna.study.StudyDirection` of the :class:`~optuna.study.Study`.
+
+            .. note::
+                This attribute is only available during single-objective optimization.
+        directions:
+            A sequence of :class:`~optuna.study.StudyDirection` objects.
+        user_attrs:
+            Dictionary that contains the attributes of the :class:`~optuna.study.Study` set with
+            :func:`optuna.study.Study.set_user_attr`.
+        system_attrs:
+            Dictionary that contains the attributes of the :class:`~optuna.study.Study` internally
+            set by Optuna.
+
+    """
```

Comment:
I agree that `FrozenStudy` sounds like it will contain a list of `FrozenTrial`. IMO, `StudyConfig` sounds like an entity user directly modify to change the configuration of `Study`, which is not true in our scenario. (Users change the config by calling some method in `Study` and `StudyConfig` is indirectly modified afterwards.) I think other than `StudySetting`, `StudyInfo`, `StudyEntry`, `StudyContext` also look like an option.

Reply:
Not a strong opinion, but I'd say it is also important to give it a similar name since the concept is similar to FrozenTrial.

Reply:
Hmm, I don't think the current `FrozenStudy` is similar to `FrozenTrial` because `FrozenTrial` can be a snapshot of corresponding object (i.e., `Trial`) but `FrozenStudy` cannot. This is mainly because `Study` has the `trials` attribute and `FrozenStudy` doesn't.

We might need to reorganize `Study` since it contains two different information:
- Experimental settings (i.e., information in `FrozenStudy`)
- Experimental results (i.e., a list of `FrozenTrial`s)
I guess it is out-of-scope of this PR, and I'd like to discuss it in the next major release (v4?).